### PR TITLE
WIP Route analysis and bugfix.

### PR DESCRIPTION
**Pull request by did-g**
Hi,
Add the ability to compute speed along an existing route.

Regards
Didier

### CODE REVIEW

**Comment by rgleason** [comment]
Didier,  This is a big important change. What a huge contributor you are!  All users will thank you!
WR Route --> Track --> Route Manager
JSON message to request Grib Values
Add SOG to route position
I am sure there are other things I have not noticed.
The plugin works well. This is an aside, there are a couple of things I have noticed in the past that are a little flakey:
- When the route isobars are being calculated, I have to refresh the screen to get it to draw, (ie zoom in or out). This is on a Windows machine.
 - In Grib_pi I set the grib date and time, then I go to wxRoute and pick "Grib Time", I find this is kind of stuck and does not change. I have not figured out what causes this, but the Compute is failing, until I get the "Grib Time" to actually work.  I will keep an eye on this.

I have some screen shots to show a little what I did. First completed a new routing.
Screenshot (71)

Weather tracks are automatically added complete with name and date.  SOW added to :!
Screenshot (72)

**Comment by rgleason** [comment]
Different Leg Speed & Arrival Dates for Track and Route.

Screenshot (77)

**Comment by seandepagnier** [comment]
wow,  I intend to  merge in a few days, give me time to review this

**Comment by rgleason** [comment]
Also this resolves the speed confusion in FR #248

**Comment by did-g** [comment]
Two things
- It doesn't solve the rendering issue on windows with wx3.1 (ie redraw is
blocked until you zoom or collapse the config dialog).

- The route analysis isn't finished, it should compute intermediate points
on grib step, I should make two PR one for fixes , one for the route stuff
but they are badly mixed.

**Comment by rgleason** [comment]
UI Improvement: Provide a yellow box "Info" tooltip at the Cursor along the Optimal Route #177

**Comment by rgleason** [comment]
Related to this improvement:
Feature Request: Grib Weather at current Optimal Route Position (option) checkbox #203

**Comment by rgleason** [comment]
I suppose this is piling onto Didier, which is unfair, but this little bug needs a fix too:
ov50 - Bug - wx-Routing requires Climatology_pi installed to be able to disable Climatology use.  #241

**Comment by did-g** [comment]
@rgleason 
about

> isochron reaching end point, set the right timespan

for computing legs speed WR is using distance/time with time being the configured step, but it doesn't work for the last leg.
Let say you have a 3h hours step but you reach the end point in 1 hour, last leg speed is distance/1h not distance/3h

**Comment by rgleason** [comment]
Ah. Ha. The light bulb just turned on. Thanks. This is very similar to the solution "Slow Route", but it is and ending or completing case, rather than starting in the middle of constrained land areas, and going through mid route land areas.

Is there some "slow route" code that could be used for begining, middle and end, when in constrained areas.

**Comment by seandepagnier** [comment]
> @rgleason
> about
> 
> > isochron reaching end point, set the right timespan
> 
> for computing legs speed WR is using distance/time with time being the configured step, but it doesn't work for the last leg.
> Let say you have a 3h hours step but you reach the end point in 1 hour, last leg speed is distance/1h not distance/3h

Do you mean the propagate to end routine?  Yes, the weather route has a special update just to reach the end position which it uses to compute the time to reach it rather than moving a fixed time step.

**Comment by rgleason** [comment]
Thank Sean, 
This seems like a very similar problem to various "Failures" at the start and middle of routes....something, some set of parameters for calculation have to be modified, in order to continue the calculation to completion (IE reaching Finish). So I ask this again...

> Is there some "slow route" code that could be used for beginning, middle and end, when in constrained areas.

**Comment by did-g** [comment]
@seandepagnier 
This fixes a display issue not a computation one.
master
bad speed
PR
right speed

**Comment by rgleason** [comment]
Please see  http://www.cruisersforum.com/forums/f134/weather-routing-100060.html#post2892375

File: src/RouteMap.cpp [hunk 1]
```diff
@@ -749,7 +749,7 @@ bool Position::Propagate(IsoRouteList &routelist, RouteMapConfiguration &configu
             return false;
     }
 
-    double timeseconds = configuration.DeltaTime;
+    double timeseconds = configuration.UsedDeltaTime;
```
Comment (seandepagnier) [inline src/RouteMap.cpp]: Why "UseDeltaTime"  Couldn't it be simplified to always specify the delta time?

**Comment by rgleason** [comment]
From my forum notes...

Didier, I tested the plugin you reference above. Oddly, the SOG and SOW do not show together as in your screenshot. Do I need to adjust a setting?

Well I turned on average speed and average speed ground. (I don't think these settings control the "Route Positon" Panel. I believe they control the parameters shown in the Configurations area.

Then I discovered several things:
1. Weather_routing seems to favor the left panel. Properly showing the isochrones as they are created. The right panel is not drawn until the route is completed.
2. Grib_pi seems to favor the right panel, showing the grib wind indications in that panel only.
3. Climatology_pi seems to do any panel that has been activated, and then the Wind Roses show.
4. When "View > Route Position" is activated and hover over Climatology sourced data/isochrones the SOG and SOW show on various isochrones, but not all of them. Some are left out, - is this because of interopolation?
5. When "View > Position" is activated and hover over Grib sourced data/isochrones the Boat Speed SOG is the only parameter that seems to show.

Can someone please confirm this and report back using Didier's appveyor "Artifact Tab" build of the plugin above?
https://ci.appveyor.com/project/sean...5223/artifacts

File: src/RouteMapOverlay.cpp [hunk 2]
```diff
@@ -1686,35 +1686,30 @@ Position* RouteMapOverlay::getClosestRoutePositionFromCursor(double cursorLat, d
...
-    Position *newPos = nullptr;
-    
-    // Get position first
-    for(Position *p = pos; p && p->parent; p = p->parent)
+    std::list<PlotData> plot = GetPlotData(false);
+    bool found = false;
+    for ( const auto &it : plot)
     {
         // Calculate distance
         // Almost like a plan (x,y) because of small distance -- is that correct?
```
Comment (seandepagnier) [inline src/RouteMapOverlay.cpp]: Is this an optimization?  Why not rebuild the plot data rather than compute the position?

**Pull request merged**